core/src/main/java/zipkin/reporter/AsyncReporter.java

```diff
@@ -0,0 +1,253 @@
+
+    static <S> BoundedAsyncReporter<S> create(Builder<S> builder, BoundedQueue<S> queue) {
+      AtomicBoolean closed = new AtomicBoolean(false);
+      CountDownLatch shutdown = new CountDownLatch(1);
+      BoundedAsyncReporter<S> result = new AutoValue_AsyncReporter_BoundedAsyncReporter<>(
+          queue,
+          builder.sender,
+          builder.metrics,
+          closed,
+          shutdown
+      );
+      new Thread(() -> {
+        try {
+          while (!closed.get()) {
+            result.flush();
+          }
+        } finally {
+          shutdown.countDown();
+        }
+      }, "Flusher(" + builder.sender + ")").start();
+      return result;
+    }
+
+    /** Returns true if the was accepted onto the queue. */
+    @Override
+    public void report(S span) {
+      metrics().incrementSpans(1);
+      if (closed().get() || !pending().offer(span)) {
+        metrics().incrementSpansDropped(1);
+      }
+    }
+
+    /**
+     * Calling this will flush any pending spans to the transport on the current thread.
+     */
+    @Override
+    public void flush() {
+      if (closed().get()) throw new IllegalStateException("closed");
+
+      List<S> drained = pending().poll(secondInNanos);
+      if (drained.isEmpty()) return;
+
+      int count = drained.size();
+      Callback callback = sendSpansCallback(count);
+      try {
+        sender().sendSpans(drained, callback);
+      } catch (RuntimeException e) {
+        callback.onError(e);
+      }
+    }
+
+    @Override public CheckResult check() {
+      return sender().check();
+    }
+
+    /**
+     * Threads block up to a second for messages in the queue. This will wait up to 3 seconds fora
+     * clean shutdown.
+     */
+    @Override
+    public void close() {
+      closed().set(true);
+      try {
+        if (!shutdown().await(3, TimeUnit.SECONDS)) {
+          logger.warning("Timed out waiting for shutdown");
+        }
+      } catch (InterruptedException e) {
+        logger.warning("Interrupted waiting for shutdown");
+      }
+      List<S> dropped = pending().drain();
+      int count = dropped.size();
+      metrics().incrementSpansDropped(count);
+      logger.warning("Dropped " + count + " spans due to shutdown");
+    }
+
+    Callback sendSpansCallback(final int count) {
+      return new Callback() {
+        @Override public void onComplete() {
+        }
+
+        @Override public void onError(Throwable t) {
+          metrics().incrementSpansDropped(count);
+          warn("Dropped " + count + " spans", t);
+        }
+      };
+    }
+
+    void warn(String message, Throwable e) {
+      message = format("%s due to %s(%s)", message, e.getClass().getSimpleName(),
+          e.getMessage() == null ? "" : e.getMessage());
+      logger.log(WARNING, message, e);
+    }
+  }
+
+  @AutoValue
+  static abstract class EncodingAsyncReporter<S, B> extends AsyncReporter<S> {
+
+    abstract Encoder<S, B> encoder();
```

Comment:
@felixbarny here's where I'd add javadoc explaining that this thing ends up controlling the element type of the bounded queue, using examples of `S` being `zipkin.Span` and `B` being `byte[]`.